src/Orleans.Runtime/Placement/Rebalancing/ActiveRebalancerGrain.cs

```diff
@@ -0,0 +1,723 @@
+                    // to take into account the edge not only from a source's prespective, but also the target's one, as it too will take part on the candidate set.
+
+                    var flippedEdge = CreateEdge(counter.Flip());
+                    return new VertexEdge[2] { edge, flippedEdge };
+                }
+
+                return new VertexEdge[1] { edge };
+            });
+
+        return result;
+
+        VertexEdge CreateEdge(EdgeCounter counter)
+        {
+            var direction = Direction.Unspecified;
+
+            direction = IsSourceThisSilo(counter)
+                ? IsTargetThisSilo(counter) ? Direction.LocalToLocal : Direction.LocalToRemote
+                : Direction.RemoteToLocal;
+
+            Debug.Assert(direction != Direction.Unspecified);   // this can only occur when both: source and target are remote (which can not happen)
+
+            return direction switch
+            {
+                Direction.LocalToLocal => new(
+                    Id: counter.Edge.Source.Id,                     // 'local' vertex was the 'source' of the communication
+                    IsMigratable: counter.Edge.Source.IsMigratable,
+                    ConnectedId: counter.Edge.Target.Id,
+                    PartnerSilo: ThisSilo,                          // the partner was 'local' (note: this.Silo = Source.Silo = Target.Silo)
+                    Direction: direction,
+                    Weight: counter.Value),
+
+                Direction.LocalToRemote => new(
+                    Id: counter.Edge.Source.Id,                     // 'local' vertex was the 'source' of the communication
+                    IsMigratable: counter.Edge.Source.IsMigratable,
+                    ConnectedId: counter.Edge.Target.Id,
+                    PartnerSilo: counter.Edge.Target.Silo,          // the partner was 'remote'
+                    Direction: direction,
+                    Weight: counter.Value),
+
+                Direction.RemoteToLocal => new(
+                    Id: counter.Edge.Target.Id,                     // 'local' vertex was the 'target' of the communication
+                    IsMigratable: counter.Edge.Target.IsMigratable,
+                    ConnectedId: counter.Edge.Source.Id,
+                    PartnerSilo: counter.Edge.Source.Silo,          // the partner was 'remote'
+                    Direction: direction,
+                    Weight: counter.Value),
+
+                _ => throw new InvalidOperationException($"The edge direction {direction} cant happen.")
+            };
+        }
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        bool IsSourceThisSilo(EdgeCounter counter) => counter.Edge.Source.Silo.IsSameLogicalSilo(ThisSilo);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        bool IsTargetThisSilo(EdgeCounter counter) => counter.Edge.Target.Silo.IsSameLogicalSilo(ThisSilo);
+    }
+
+    private enum Direction
```

Comment:
```suggestion
    private enum Direction : byte
```
To make the enum smaller (1 byte vs. 4 bytes (int is default type)).